src/ray/gcs/tables.cc

```diff
@@ -373,55 +373,138 @@ void ClientTable::RegisterClientRemovedCallback(const ClientTableCallback &callb
   client_removed_callback_ = callback;
   // Call the callback for any removed clients that are cached.
   for (const auto &entry : client_cache_) {
-    if (!entry.first.is_nil() && !entry.second.is_insertion) {
+    if (!entry.first.is_nil() && (entry.second.entry_type == EntryType::DELETION)) {
       client_removed_callback_(client_, entry.first, entry.second);
     }
   }
 }
 
+void ClientTable::RegisterResourceCreateUpdatedCallback(
+    const ClientTableCallback &callback) {
+  resource_createupdated_callback_ = callback;
+  // Call the callback for any clients that are cached.
+  for (const auto &entry : client_cache_) {
+    if (!entry.first.is_nil() &&
+        (entry.second.entry_type == EntryType::RES_CREATEUPDATE)) {
+      resource_createupdated_callback_(client_, entry.first, entry.second);
+    }
+  }
+}
+
+void ClientTable::RegisterResourceDeletedCallback(const ClientTableCallback &callback) {
+  resource_deleted_callback_ = callback;
+  // Call the callback for any clients that are cached.
+  for (const auto &entry : client_cache_) {
+    if (!entry.first.is_nil() && (entry.second.entry_type == EntryType::RES_DELETE)) {
```

Comment:
The parenthesis seems not necessary. `(entry.second.entry_type == EntryType::RES_DELETE)`->`entry.second.entry_type == EntryType::RES_DELETE`.